### PR TITLE
Better ns support.

### LINKED ISSUE #11: allow binding the control process to the router

### LINKED ISSUE #14: support dynamic topologies
a machine should be able to move to a different network

### PR DESCRIPTION
**Pull request by dvc94ch**
Maybe this is a better approach for #13 

TODO
- [x] revisit spawn_machine/spawn_network api
- [x] use !Send guard for namespace
- [x] write ping test
- [x] write enter namespace test
- [x] write test for changing topology

Closes #11 
Closes #14

### CODE REVIEW

**Review by wngr — COMMENTED** [review]
So usage would look like this?
```rust
let current = Namespace::current()?;
network.machines[0].namespace.enter()?;
// do stuff
current.enter()?; // returns to the old namespace
```

I like that. Maybe this could be made more ergonomic like so:
```rust
let guard = network.machines[0].namespace.enter()?;
// do stuff
drop(guard); // returns to the old namespace
```

WDYT?

**Comment by wngr** [comment]
That means though, that any futures need to be run on the current thread and not in some thread pool. Not sure whether we should offer a nicer API for that.. The API proposed in #13 took care of that.

**Comment by dvc94ch** [comment]
we can make the guard !Send right? that should make sure that it stays in the current thread.

**Comment by wngr** [comment]
You could still do something like
```rust
let guard = ..;
my_fut.await;
drop(guard)
```
and `my_fut` is executed on some other thread. Might just be a better of documentation though.

**Comment by dvc94ch** [comment]
So implementing !Send is unstable, didn't remember. So instead you could define a `NamespaceGuard(*mut std::ffi::c_void)`. However I'm not sure if that is a good idea after all, because `Namespace::current` gives the namespace of the thread and then you enter another threads namespace. So these guards don't really nest and it should be fairly straight forward to go back to the current threads namespace using `Namespace::current()?.enter()?`.

File: src/lib.rs [hunk 1]
```diff
@@ -16,88 +15,184 @@ pub fn run<F>(f: F)
...
 
-    pub fn addr(&self) -> Ipv4Addr {
-        self.addr
+pub struct Netsim<C, E> {
+    machines: Vec<Machine<C, E>>,
+    plugs: Vec<Connector>,
+    networks: Vec<Network>,
+}
+
+impl<C, E> Default for Netsim<C, E> {
```
Comment (wngr) [inline src/lib.rs:43]: replace with `#[derive(Default)]`?

File: router/src/lib.rs [hunk 2]
```diff
@@ -1,90 +1,109 @@
...
-                    Poll::Ready(None) => {
-                        self.rxs.remove(i);
-                        self.txs.remove(i);
-                        break;
+                    Some(RouterCtrl::RemoveRoute(id, ch)) => {
+                        let plug = if let Some(idx) = conns.iter().position(|(id2, _, _)| *id2 == id) {
+                            let (_, plug, _) = conns.swap_remove(idx);
+                            Some(plug)
+                        } else {
+                            None
```
Comment (wngr) [inline router/src/lib.rs:57]: When does this return None?

**Review by wngr — COMMENTED** [review]
The new API looks great. I will give this a spin ~tomorrow~.

File: machine/src/lib.rs [hunk 3]
```diff
@@ -12,81 +12,173 @@ macro_rules! errno {
...
+        self.mask
+    }
+
+    pub fn set_addr(&mut self, addr: Ipv4Addr, mask: u8) {
+        self.ctrl
+            .unbounded_send(IfaceCtrl::SetAddr(addr, mask))
+            .unwrap();
+        self.addr = addr;
+        self.mask = mask;
+    }
```
Comment (wngr) [inline machine/src/lib.rs]: This is potentially racy, isn't it. Maybe make it async and wait for `IfaceCtrl` to ack?

File: src/lib.rs (hunk 1 above)
Comment (dvc94ch) [reply src/lib.rs:43]: That requires a Default trait bound on C and E which are probably going to be enums

File: router/src/lib.rs (hunk 2 above)
Comment (dvc94ch) [reply router/src/lib.rs:57]: if there is no route to `id`, which could be for example because the machine exited.

File: machine/src/lib.rs (hunk 3 above)
Comment (dvc94ch) [reply machine/src/lib.rs]: fixed

**Review by wngr — APPROVED** [review]
In the meantime I also gave the API a try :+1: 

Docs would be nice :trollface:

**Pull request merged**